googledataprocauthenticator/magics/dataprocmagics.py

```diff
@@ -0,0 +1,238 @@
+           sessions
+               Display the available Livy sessions and other configurations for sessions created with Google Authentication.
+           add
+               Add a Livy session given a session name (-s), language (-l), and endpoint credentials.
+               The -k argument, if present, will skip adding this session if it already exists.
+               e.g. `%spark add -s test -l python -u https://sparkcluster.net/livy -t Kerberos -a u -p -k`
+           config
+               Override the livy session properties sent to Livy on session creation. All session creations will
+               contain these config settings from then on.
+               Expected value is a JSON key-value string to be sent as part of the Request Body for the POST /sessions
+               endpoint in Livy.
+               e.g. `%%spark config`
+                    `{"driverMemory":"1000M", "executorCores":4}`
+           run
+               Run Spark code against a session.
+               e.g. `%%spark -s testsession` will execute the cell code against the testsession previously created
+               e.g. `%%spark -s testsession -c sql` will execute the SQL code against the testsession previously created
+               e.g. `%%spark -s testsession -c sql -o my_var` will execute the SQL code against the testsession
+                        previously created and store the pandas dataframe created in the my_var variable in the
+                        Python environment.
+           logs
+               Returns the logs for a given session.
+               e.g. `%spark logs -s testsession` will return the logs for the testsession previously created
+           delete
+               Delete a Livy session.
+               e.g. `%spark delete -s defaultlivy`
+           cleanup
+               Delete all Livy sessions created by the notebook. No arguments required.
+               e.g. `%spark cleanup`
+        """
+        usage = "Please look at usage of %spark by executing `%spark?`."
+        user_input = line
+        args = parse_argstring_or_throw(self.spark, user_input)
+        subcommand = args.command[0].lower()
+        if args.auth == "Google" and subcommand == "add":
+            if args.url is None:
+                self.ipython_display.send_error("Need to supply URL argument (e.g. -u https://example.com/livyendpoint)")
+                return
+            name = args.session
+            language = args.language
+            endpoint = Endpoint(args.url, initialize_auth(args))
+            self.endpoints[args.url] = endpoint
+            # get current stored_endpoints
+            stored_endpoints = self.get_stored_endpoints()
+            endpoint_tuple = (args.url, endpoint.auth.active_credentials)
+            stored_endpoints.append(endpoint_tuple)
+            # stored updated stored_endpoints
+            self.db['autorestore/' + 'stored_endpoints'] = stored_endpoints
```

Comment:
If we do this, won't we end up with an ever-growing list of endpoints because we don't ever remove from the list? I think it's better to save the current set of endpoints that the spark controller has.

Reply:
I added this. So now it will check to see if args.url already exists as a key in self.endpoints dict (which is loaded from get_stored_endpoints()) https://github.com/GoogleCloudDataproc/dataprocmagic/pull/3/files#diff-28b0b312480d3bf4199e359119077b98R208. If it doesn't exist, then it will add the endpoint tuple to stored endpoints. This is necessary because each time the extension is reloaded the sparkcontroller's endpoints are all cleared out.